cmd/govim/highlight.go

```diff
@@ -86,18 +105,116 @@ func (v *vimstate) redefineHighlights(diags []types.Diagnostic, force bool) erro
 		v.BatchChannelCall("prop_add",
 			d.Range.Start.Line(),
 			d.Range.Start.Col(),
-			struct {
-				Type    string `json:"type"`
-				EndLine int    `json:"end_lnum"`
-				EndCol  int    `json:"end_col"` // column just after the text
-				BufNr   int    `json:"bufnr"`
-			}{string(hi), d.Range.End.Line(), d.Range.End.Col(), d.Buf})
+			propAddDict{string(hi), types.DiagnosticTextPropID, d.Range.End.Line(), d.Range.End.Col(), d.Buf},
+		)
+	}
+
+	v.BatchEnd()
+	return nil
+}
+
+func (v *vimstate) updateReferenceHighlight(refresh bool) error {
+	if v.config.HighlightReferences == nil || !*v.config.HighlightReferences {
+		return nil
+	}
+	b, pos, err := v.cursorPos()
+	if err != nil {
+		return fmt.Errorf("failed to get current position: %v", err)
+	}
+
+	// refresh indicates if govim should call DocumentHighlight to refresh
+	// ranges from gopls since we want to refresh when the user goes idle,
+	// and remove highlights as soon as the user is busy. To prevent
+	// flickering we keep track of the current highlight range and avoid
+	// removing text properties if the cursor is still within the range.
+	if !refresh {
+		if v.currentReference != nil && pos.IsWithin(*v.currentReference) {
+			return nil
+		}
+		v.currentReference = nil
+		v.removeTextProps(types.ReferencesTextPropID)
+		return nil
+	}
+
+	// Cancel any ongoing requests to make sure that we only process the
+	// latest response.
+	if v.cancelDocHighlight != nil {
+		v.cancelDocHighlight()
+		v.cancelDocHighlight = nil
+	}
+	ctx, cancel := context.WithCancel(context.Background())
+	v.cancelDocHighlight = cancel
+
+	v.tomb.Go(func() error {
+		v.redefineReferenceHighlight(ctx, b, pos)
+		cancel()
```

Comment:
For the sake of consistency with the code and comment above, this should probably be:

```
v.cancelDocHighlight()
```